### PR TITLE
[Fusilli] Generalize physical tensor layout support

### PR DESCRIPTION
**Pull request by IanWood1**
Currently, Fusilli only supports 2 physical tensor layouts: NCHW (contiguous) and NHWC. However, pointwise ops and matmul ops (https://github.com/nod-ai/shark-ai/pull/2632) require support for any layout that can be represented as a permutation of dimensions.

#### 1. Added new validation method `hasValidPhysicalRepresentation()`
This method checks that the combination of stride & logical dimensions can be represented as a physical tensor and that the logical tensor is a transpose of the physical tensor.

#### 2. Generalized `getPhysicalDim()`
This now supports getting the physical dims for when they are a permutation of the logical dims. This uses `getLogicalToPhysicalPerm()` to get the physical dims by transposing the logical dims.

#### 3. Updated the ASM emitters
Updated the ASM emitters to use the new helper `getLogicalToPhysicalPerm()`. This is the same utility used by `getPhysicalDims()`, which ensures that the transpose vec matches the expected type. This is mostly an NFC and provides a cleaner and more concise way to get the permutation.

Also, pointwise nodes weren't converting the physical tensor to a logical tensor, which was incorrect. This PR breaks pointwise ops without fixing this, so the fix for pointwise ops is included here.

#### 4. Updated CHECK lines in conv tests
`getPhysicalDim` assumed that physical memory would be stored only in NCHW or NHWC format. That meant there was a unique way to go from logical to physical. Now that this is generalized, we run into a problem: unit dims add ambiguity and so there is no longer a unique physical tensor representation of the logical dims. Unit dims can be placed anywhere without changing the meaning of the tensor. However, we still need a single unique representation for the MLIR types to match. When finding the physical tensor, this PR chooses to place the non-unit dims in the same location as the logical tensor. This provides a way to find a unique physical representation of the tensor.

For example:

```
k = 256, c = 128, r = 1, s = 1
dim = [k, c, r, s] = [256, 128, 1, 1]
stride = [c * r * s, r * s, s, 1] = [128, 1, 1, 1]
```

Previously, this would result in a physical tensor of `!torch.vtensor<[256,1,1,128],f32>` but now this results in `!torch.vtensor<[256,128,1,1],f32>`.

### CODE REVIEW

**Comment by IanWood1** [comment]
TODOs
- Use hasValidPhysicalRepresentation to verify nodes (conv already checks layout which is a subset of hasValidPhysicalRepresentation)

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 1]
```diff
@@ -257,6 +257,62 @@ getContiguousToChannelsLastPermuteOrder(size_t numDims) {
...
+    inverse[permutation[i]] = static_cast<int64_t>(i);
+  }
+  return inverse;
+}
+
+// This computes the permutation needed to go from logical dims to physical dims
+// based on the dims and stride. Unit length dimensions are kept in their
+// original positions as they don't meaningfully contribute to the layout.
+inline std::vector<int64_t>
+getLogicalToPhysicalPerm(const std::vector<int64_t> &dim,
```
Comment (sjain-stanford) [inline sharkfuser/include/fusilli/attributes/tensor_attributes.h]:
```suggestion
getLogicalToPhysicalPermuteOrder(const std::vector<int64_t> &dim,
```
for consistency with similar utilities that return the permute order

File: sharkfuser/tests/lit/test_conv_asm_emitter_nchw_krsc.cpp [hunk 2]
```diff
@@ -13,7 +13,7 @@
 // clang-format off
 //
 // TORCH-CHECK:   module @module {
-// TORCH-CHECK:     func.func @main(%result_: !torch.tensor<[16,256,64,32],f32>, %arg0_image: !torch.vtensor<[16,128,64,32],f32>, %arg1_filter: !torch.vtensor<[256,1,1,128],f32>) attributes {torch.assume_strict_symbolic_shapes} {
+// TORCH-CHECK:     func.func @main(%result_: !torch.tensor<[16,256,64,32],f32>, %arg0_image: !torch.vtensor<[16,128,64,32],f32>, %arg1_filter: !torch.vtensor<[256,128,1,1],f32>) attributes {torch.assume_strict_symbolic_shapes} {
```
[the remark this answers is not in the document]
Comment (IanWood1) [reply sharkfuser/tests/lit/test_conv_asm_emitter_nchw_krsc.cpp]: No, this comes from the change to `getPhysicalDim()`. Note that MLIR that you linked to is from the Graph node, not the conv node (since it is the function arg). The graph node should have no concept of NCHW/NHWC. It just determines the physical tensor from the logical tensor + permutation obtained from the stride. The changes to the tests are specific to the ambiguity that unit dims introduce. I.e., no changes were made to tests that had no unit dims. From the PR description:

>Now that this is generalized, we run into a problem: unit dims add ambiguity and so there is no longer a unique physical tensor representation of the logical dims. Unit dims can be placed anywhere without changing the meaning of the tensor. However, we still need a single unique representation for the MLIR types to match. When finding the physical tensor, this PR chooses to place the non-unit dims in the same location as the logical tensor.

File: sharkfuser/tests/test_tensor_attributes.cpp [hunk 3]
```diff
@@ -265,6 +265,130 @@ TEST_CASE("TensorAttr isContiguous and isChannelsLast checks", "[TensorAttr]") {
   REQUIRE(t2.getPhysicalDim() == std::vector<int64_t>{2, 4, 3});
 }
 
+TEST_CASE("TensorAttr getPhysicalDim", "[TensorAttr]") {
```
Comment (sjain-stanford) [inline sharkfuser/tests/test_tensor_attributes.cpp]: Also needs unit tests for `getLogicalToPhysicalPerm`.

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 4]
```diff
@@ -467,21 +523,65 @@ class TensorAttr {
...
+  }
+
+  // Convert logical dims + stride into physical dims. The stride
+  // represents how dimensions are accessed in physical memory.
+  //
+  // Dimensions of size 1 don't meaningfully contribute to the physical layout
+  // and are kept in their relative positions to preserve information.
+  std::vector<int64_t> getPhysicalDim() const {
+    assert(hasValidPhysicalRepresentation() &&
+           "Tensor has invalid physical representation");
```
[the remark this answers is not in the document]
Comment (IanWood1) [reply sharkfuser/include/fusilli/attributes/tensor_attributes.h]: Added this, and now `SECTION("Zero dimension tensor")` (at the top of the tensor tests) checks for failure instead.

**Review by sjain-stanford — APPROVED** [review]
LGTM, please feel free to land after adding the sample.

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 5]
```diff
@@ -467,21 +548,65 @@ class TensorAttr {
...
+    // and the expected product of dimensions should match
+    int64_t expectedStride = 1;
+    for (size_t i = 0; i < numDims; ++i) {
+      int64_t actualStride = strideAndDim[i].first;
+      int64_t dimSize = strideAndDim[i].second;
+
+      // For unit length dimensions, any stride value is technically valid
+      if (dimSize == 1) {
+        continue;
+      }
```
Comment (sjain-stanford) [inline sharkfuser/include/fusilli/attributes/tensor_attributes.h]: nit: remove braces for single line ifs

File: sharkfuser/include/fusilli/support/asm_emitter.h [hunk 6]
```diff
@@ -1090,27 +1052,98 @@ inline std::string ConvDGradNode::emitNodePreAsm() const {
...
+  // Get the input tensor based on index
+  std::shared_ptr<TensorAttr> inputT = pointwiseAttr.inputs.at(
+      static_cast<PointwiseAttr::InputNames>(inputIndex));
+  std::ostringstream oss;
+  std::string prefix = "permute_IN_" + std::to_string(inputIndex);
+  std::string suffix = pointwiseAttr.getName();
+
+  // Emit permute dimensions based on layout.
+  oss << getListOfIntOpsAsm(
+      getLogicalToPhysicalPermuteOrder(inputT->getDim(), inputT->getStride()),
```
Comment (sjain-stanford) [inline sharkfuser/include/fusilli/support/asm_emitter.h]: I'm a little confused by the naming of this. Aren't we going from physical -> logical on the inputs? 

Consider operand `bT` (IN_1) of logical dim `{128, 256}` stride `{1, 128}`.
```c++
  auto bT =
      graph->tensor(TensorAttr()
                        .setName("arg1_add_transposed")
                        .setDim({128, 256})
                        .setStride({1, 128})); // Transposed: stride={1, 128}
```

In the emitted MLIR, transpose ops are inserted to permute operand `bT` from physical dims `{256, 128}` to logical dims `{128, 256}` before applying the pointwise op, right?

File: sharkfuser/include/fusilli/support/asm_emitter.h [hunk 7]
```diff
@@ -477,14 +469,11 @@ inline std::string ConvFPropNode::getPermuteYOpsAsm() const {
...
-    oss << getListOfIntOpsAsm(
-        getPreserveContiguousPermuteOrder(yT->getDim().size()), prefix, suffix);
-  else
-    oss << getListOfIntOpsAsm(
-        getContiguousToChannelsLastPermuteOrder(yT->getDim().size()), prefix,
-        suffix);
+  // Emit permute dimensions based on layout (inverse to convert back from
+  // contiguous).
+  oss << getListOfIntOpsAsm(inversePermutation(getLogicalToPhysicalPermuteOrder(
+                                yT->getDim(), yT->getStride())),
```
Comment (sjain-stanford) [inline sharkfuser/include/fusilli/support/asm_emitter.h]: For readability, consider creating a separate utility `getPhysicalToLogicalPermuteOrder` that internally calls `inversePermutation` on the `getLogicalToPhysicalPermuteOrder`...  

(btw these will get swapped after addressing the naming comment..)

File: sharkfuser/include/fusilli/support/asm_emitter.h (hunk 6 above)
Comment (sjain-stanford) [reply sharkfuser/include/fusilli/support/asm_emitter.h]: This should be called `getPhysicalToLogicalPermuteOrder` instead ([comment](https://github.com/nod-ai/shark-ai/pull/2653#discussion_r2512981432)). It makes sense because we want the ops to operate on logical dims but the main inputs/outputs are physical, so the translation from inputs -> op is physical->logical and from op -> outputs is logical -> physical (inverse).

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 8]
```diff
@@ -467,21 +548,65 @@ class TensorAttr {
...
+  // Check if the stride pattern is valid and can represent a physical tensor.
+  // A valid stride pattern must satisfy:
+  // 1. Each stride is a product of dimensions in faster-changing positions
+  // 2. The strides are consistent with some permutation of dimensions
+  bool hasValidPhysicalRepresentation() const {
     size_t numDims = dim_.size();
+    if (numDims == 0)
+      return true;
+    if (numDims != stride_.size())
+      return false;
```
Comment (a-sidorova) [inline sharkfuser/include/fusilli/attributes/tensor_attributes.h]: Could you please move this check up? It seems like it should be before empty tensor validation

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 9]
```diff
@@ -467,24 +486,153 @@ class TensorAttr {
...
+      size_t targetIdx = strideWithIndex[i].second;
+      permuteOrder[logicalIdx] = static_cast<int64_t>(targetIdx);
+    }
+
+    return permuteOrder;
+  }
+
+  // This computes the permutation needed to go from physical dims to logical
+  // dims. This is the inverse of getLogicalToPhysicalPermuteOrder().
+  std::vector<int64_t> getPhysicalToLogicalPermuteOrder() const {
```
Comment (sjain-stanford) [inline sharkfuser/include/fusilli/attributes/tensor_attributes.h]: nit: please unit test this

File: sharkfuser/include/fusilli/attributes/tensor_attributes.h [hunk 10]
```diff
@@ -257,6 +257,82 @@ getContiguousToChannelsLastPermuteOrder(size_t numDims) {
...
+// This computes the permutation needed to go from logical dims to physical dims
+// based on the dims and stride. Unit length dimensions are kept in their
+// original positions as they don't meaningfully contribute to the layout.
+//
+// Note: perm[i] means that the i-th dimension in the logical layout is mapped
+// to the perm[i]-th dimension in the physical layout.
+//
+// Examples:
+//   1. Contiguous NCHW layout: dim={2, 3, 4}, stride={12, 4, 1}
+//      Returns: {0, 1, 2} (identity - already in physical order)
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply sharkfuser/include/fusilli/attributes/tensor_attributes.h]: Oh I see, thank you for the explanation!

File: sharkfuser/samples/pointwise/pointwise_add_transposed.cpp [hunk 11]
```diff
@@ -20,18 +20,20 @@
...
   };
 
   // Result of inputData + transpose(inputData)
   const std::vector<float> expectedResult = {
-    2.0f, 5.0f,
-    5.0f, 8.0f
+    2.0f, 6.0f,
+    5.0f, 9.0f,
+    8.0f, 12.0f
   };
```
[the remark this answers is not in the document]
Comment (IanWood1) [reply sharkfuser/samples/pointwise/pointwise_add_transposed.cpp]: The current implementation uses the stride of the first input that matches the output shape. This matches Cudnn's frontend. However, if users want a specific output stride, they should specify it explicitly.
